### DISCUSSION TITLE
Example Tailwind regex breaks if parentheses are used in comments

### DISCUSSION THREAD

**Discussion by pm0u** [joe-bell/cva#85] [Q&A] [answered]
**Describe the bug**
Using the regex provided in the readme for tailwind intellisense, the intellisense breaks if a comment inside the CVA object contains a parentheses.

**To Reproduce**

1. Create a CVA object
```js
const button = cva(["p-4", {
  variants: {
    variant: {
      primary: ["bg-red-10"]
    }
  }
})
```
2. Add a comment that contains a parentheses
```js
const button = cva(["p-4", {
  variants: {
    /** Button variant (background color) */
    variant: {
      primary: ["bg-red-10"]
    }
  }
})
```
3. Observe intellisense stops working

**Expected behavior**
Intellisense would not be impacted by content in comments.

**Screenshots**

<img width="705" alt="image" src="https://user-images.githubusercontent.com/410558/203607331-9c06b278-dbd1-448e-8f44-d7ceade924a6.png">
<img width="637" alt="image" src="https://user-images.githubusercontent.com/410558/203607374-78d1b19d-cb1b-4ceb-9d46-d01ffa9d6f1c.png">

**Desktop (please complete the following information):**

- OS: MacOS
- Browser NA
- Version `tailwind@3.2.4, cva@0.3.0`

**Comment by pm0u** [comment]
The use case for us is explanatory comments that are parsed into the Storybook argsTable (in addition to dev usefulness)

**Comment by pm0u** [comment]
going to try this one out for awhile:
```json
    ["cva\\(([\\s\\S]*?)\\}[\\s]*?\\)", "[\"'`]([^\"'`]*).*?[\"'`]"]
```

Basically, instead of matching characters excluding end parentheses (`[^)]*`), match all characters (including newlines) up until a `}` character followed by any amount of whitespace which is then followed by a closing parentheses. Essentially looking for the closing `})` on the CVA object as the second param. Would break if there was a comment or something else other than whitespace between those two characters (newline acceptable). edit: had to make the `[\s\S]*` lazy otherwise it matches everything after `cva(` 😅

**Comment by pm0u** [comment]
This seems to be working well so far. The only thing I could see being an improvement is if the regex could simply "exclude" anything in a comment (either lines starting with `//` or anything between `/* */`). This could allow explaining usage with JS syntax in comments which my current suggestion does not permit (IE a comment such as `/** pass in args like function({ arg1 }) */`). I'm not quite good enough at negative lookups/regex to put this together.

**Comment by arpadgabor** [comment]
I've found this to be an issue with generics as well when declaring variants. I've been using this regex with some success but could be improved:

```
["cva<.*>\\(([^)]*)\\)", "[\"'`]([^\"'`]*).*?[\"'`]"]
```

and it works with a snippet like

```
const style = cva<Variants>(...)
```

**Comment by joe-bell** [comment]
@arpadgabor `cva` does not support nor encourage passing generic types — it’s not documented because it’s internal only

**Comment by leeschmidt4** [comment]
The tailwind regex is brittle. It breaks in a lot of strange scenarios with no clear reason. Is this a bug with vscode or tailwind? I can't see what would be wrong with the regex that causes these weird issues.

Playing around with it more, definitely seems to be tied to the length or size of the object.

**Comment by leeschmidt4** [comment]
I think this is most definitely a tailwind or vscode bug. Here's what I've done as a temporary workaround for objects that are too large and break the regex intellisense:

I wrap every class or list of classes in tw() which is defined as: `const tw = <T,>(a: T): T => a` (or set the type to ClassValue). The only change to the regex is to replace the leading `cva` with `tw`, so you have `["tw\\(([^)]*)\\)", "[\\\"'`]([^\\\"'`]*).*?[\\\"'`]"]`

It's kinda ugly, but I almost like it due to the fact it's a little more clear to whomever is reading the code that these are specifically tailwind classes.

Example:

`size: {
        small: tw(['text-sm']),
        medium: tw(['text-base']),
      },`

**Comment by corydeppen** [comment]
This looks similar to an issue I'm seeing, where intellisense breaks when using `calc` for an arbitrary value (e.g. `"w-[calc(2*2rem)]"`).

**Comment by joe-bell** [comment]
Linking @its-monotype's issue which feels related (https://github.com/joe-bell/cva/issues/108)

---

**Describe the bug**
IntelliSense does not work in `compoundVariants` despite the first element of the array in VSCode.

**To Reproduce**
Steps to reproduce the behavior:
1. Complete each step described in the README in "Installation -> TailwindCSS -> IntelliSense -> Visual Studio Code" section.
2. Try to use IntelliSense in the `compoundVariants` in any object despite the first one.

**Expected behavior**
IntelliSense work as expected as here:
![image](https://user-images.githubusercontent.com/79363260/216441025-5adeacd0-ffb2-47fb-8046-4aa953f2750c.png)


**Screenshots**
![image](https://user-images.githubusercontent.com/79363260/216439477-a9f9a9b5-f885-4885-8a97-3ccd06cf65fd.png)

**Reply by pm0u** [reply]
I wonder if this might actually be related to the length of the match, I remember seeing something in the tailwind intellisense related to that, there is a limit somewhere

**Reply by pm0u** [reply]
This was the issue/comment: https://github.com/tailwindlabs/tailwindcss/issues/7553#issuecomment-922237989

**Reply by pm0u** [reply]
Looks like that file has changed a lot - but I think this is the relevant part, it seems to now be 1000 (or 2000?) characters:

```
  const searchRange: Range = {
    start: document.positionAt(Math.max(0, positionOffset - 1000)),
    end: document.positionAt(positionOffset + 1000),
  }
```

https://github.com/tailwindlabs/tailwindcss-intellisense/blob/master/packages/tailwindcss-language-service/src/completionProvider.ts#L487-L490

**Reply by its-monotype** [reply]
I think you are right and its a limitation to 1000 characters because when I select all my cva block:
![image](https://user-images.githubusercontent.com/79363260/216447708-9e28bf0b-e4d2-49e9-abd4-b6471b8c68f6.png)

**Reply by its-monotype** [reply]
Also, look at this, maybe it's something around 1000:

https://user-images.githubusercontent.com/79363260/216448886-a75aa988-61ec-4c2d-b5de-372fa4767fde.mp4

**Reply by its-monotype** [reply]
It's a shame if it can't be fixed.

**Reply by joe-bell** [reply]
Oh shoot, nice find

**Reply by pm0u** [reply]
@its-monotype I don't think you'd be able to change that in the extension, no... but you could break up large CVA objects like that and create your own regex to match something like `const compoundVariants = { /* match strings in here */}`

**Reply by its-monotype** [reply]
Of course, the problem is unpleasant, but it prompted me to reconsider the logic of the component and it seems to me that I made it better by removing the size variant since it can be customized so easy from the outside using className prop 😅

**Comment by luke-layerhealth** [comment]
@joe-bell the PR I posted #99 resolves these parentheses issues. You can take the regex from there and it should solve most of the parentheses issues.

Here are the regexes in case its helpful for someone else. Enables standalone cx as well as cva completion.

```
    // enable autocomplete in cx(...)
    ["cx\\(((?:[^()]|\\([^()]*\\))*)\\)", "[\"'`]([^\"'`]*).*?[\"'`]"],
    // enable autocomplete in cva(...)
    ["cva\\(((?:[^()]|\\([^()]*\\))*)\\)", "[\"'`]([^\"'`]*).*?[\"'`]"]
```

**Reply by joe-bell** [reply]
Seems to work well! Unfortunately it seems to trigger IntelliSense for variant keys though

**Reply by its-monotype** [reply]
> Seems to work well! Unfortunately it seems to trigger IntelliSense for variant keys though

I don't have such a problem, everything seems to be working perfectly, how can I check this?

**Reply by luke-layerhealth** [reply]
@joe-bell also not sure what you mean. If you could post a screenshot of the issue I can try to help resolve.

**Reply by joe-bell** [reply]
No problem! As you can see here, IntelliSense is being triggered for all object values **and** keys:

<img width="498" alt="CleanShot 2023-02-21 at 13 48 10@2x" src="https://user-images.githubusercontent.com/7349341/220268441-6908d7ee-c45a-4008-bd9e-fdae5f2b5c7d.png">

<img width="563" alt="CleanShot 2023-02-21 at 13 46 01@2x" src="https://user-images.githubusercontent.com/7349341/220268320-54b55732-2a0a-4184-a8f8-a89bdfede175.png">

<img width="550" alt="CleanShot 2023-02-21 at 13 48 35@2x" src="https://user-images.githubusercontent.com/7349341/220268585-e954120e-84b1-47f2-a1e1-58308dc7e3d1.png">

Ideally, IntelliSense should be limited to:

1. The first `cva` parameter
2. Object values within `variants`, and `class`/`className` within `compoundVariants`

I have some changes in mind for v1 that should make this regex easier 🤞 

---

Really appreciate your efforts to dig into this by the way 🙏

**Reply by luke-layerhealth** [reply]
The issue is clear but haven't come up with a fix 🤷

**Reply by faergeek** [reply]
I'm currently trying out this pair of regexes:

```json
["(?:clsx|cn|cva|cx)\\(([^()]*(?:\\([^)]*\\)[^()]*)*)\\)", "[\"'`]([^\"'`]*)[\"'`]"]
```

Seems to work fine. It's similar to the one at the start of this thread, but groups different functions in a non-capturing group in a single entry and I've also removed `.*?` from class regex as it doesn't seem to make much sense to have it (why exclude quotes at first and then just match anything including quotes after that anyway?).

As for triggering for keys, the original regex from [docs](https://cva.style/docs/getting-started/installation) has the same problem as there is no way to avoid matching keys. These are just regexes, to distinguish between a key and a value you would need a real parser: https://github.com/tailwindlabs/tailwindcss-intellisense/blob/4d4ca278be69e8496662b2b98e3ce431bd04d9f9/packages/tailwindcss-language-service/src/util/find.ts#L143-L179

**Comment by franzwollang** [comment]
My colleague is using this library and was having some trouble with Tailwind Intellisense as well. I've taken a stab at a less brittle regex, hopefully it's helpful to others: regexr.com/78lt3

My colleague is still having issues even after all of the regex tinkering but as far as we can tell, the regex isn't the primary issue. There may be something else going on with either VSCode or the Tailwind Intellisense plugin that makes it not work so well with longer CVA objects?

**Comment by deltasierra96** [comment]
Any updates on this issue? It's clearly a Regex pattern problem. I cannot for the life of me to cleanly get autocomplete to work on cva components that have compound variants.

**Reply by joe-bell** [reply]
PRs are welcome, I'm only human and I'm trying my best to work on this project with $0 MRR 🥲

**Reply by pm0u** [reply]
@deltasierra96 have you tried the suggestions in this discussion and reviewed the issues others have had? Mine is working for me in compound variants without issue.

```
    ["cva\\(([\\s\\S]+?)\\}[\\s]*?\\)", "[\"'`]([^\"'`]*).*?[\"'`]"]
```
    
    
<img width="1052" alt="image" src="https://github.com/joe-bell/cva/assets/410558/5cdce6f3-11d0-4dcf-a0aa-5722ae21eaed">

provide more context if you are actually looking for help here.

**Comment by joe-bell** [comment, marked as answer]
Should be fixed with https://github.com/joe-bell/cva/pull/332
